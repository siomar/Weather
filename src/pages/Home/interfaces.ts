export interface IWeather {
  id: number;
  main: string;
  description: string;
  icon: string;
}
export interface ILocation {
  coords: ICoords;
}

export interface ICoords {
  latitude: string;
  longitude: string;
}

export interface IAddress {
  city: string;
  country: string;
}

export interface IResponse {
  data: IData;
}

export interface IData {
  name: string;
  visibility: number;
  main: {
    humidity: number;
    temp: number;
    temp_min: number;
    temp_max: number;
  };
  sys: {
    country: string;
  };
  weather: IWeather[];
}
