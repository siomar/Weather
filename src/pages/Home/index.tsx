import React, {useEffect, useState} from 'react';
import {Alert} from 'react-native';
import api from '../../services/api';

import Geolocation from '../../utils/geolocation';

import {
  Background,
  Box,
  Title,
  Icon,
  Description,
  SafeArea,
  Details,
  Address,
  Content,
  Update,
  UpdateText,
  Table,
  Line,
  LineTitle,
  LineDecription,
  Loading,
  Temp,
  TempTitle,
} from './styles';

import bg from '../../assets/bg.jpg';
import {IAddress, ILocation, IResponse, IWeather} from './interfaces';

const Home = () => {
  const [weather, setWeather] = useState({} as IWeather);
  const [address, setAddress] = useState({} as IAddress);
  const [info, setInfo] = useState({} as any);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    getLocation();
  }, []);

  const getLocation = async () => {
    try {
      const key = 'cdc9309860b9fb1a8b1e0e8a93ecf9bf';
      const {coords}: ILocation = await Geolocation.getLocationLatLng();

      const {data}: IResponse = await api.get(
        `weather?lat=${coords.latitude}&lon=${coords.longitude}&lang=pt&appid=${key}&units=metric`,
      );
      setWeather(data.weather[0]);
      setAddress({
        city: data.name,
        country: data.sys.country,
      });
      setInfo({
        visibility: data.visibility,
        humidity: data.main.humidity,
        temp: data.main.temp,
        temp_min: data.main.temp_min,
        temp_max: data.main.temp_max,
      });
      setLoading(false);
    } catch (error) {
      Alert.alert('Ops, ocorreu um erro!', error.message);
    }
  };

  const editTemp = (temp: number) => temp.toFixed(0);

  const updateScreen = () => {
    setLoading(true);
    getLocation();
  };
  return (
    <>
      {loading && (
        <Loading>
          <Description>Carregando...</Description>
        </Loading>
      )}

      <Background nestedScrollEnabled={true}>
        <Box source={bg} resizeMode="cover">
          <SafeArea>
            {weather && (
              <Icon
                source={{
                  uri: `https://openweathermap.org/img/wn/${weather.icon}@2x.png`,
                }}
              />
            )}
            <Title>{info.temp && editTemp(info.temp)}°C</Title>
            <Description>{weather.description}</Description>
          </SafeArea>
        </Box>
        <Content>
          <Details>
            <Temp>
              <TempTitle>{weather.main}</TempTitle>
              <Address>
                {address.city} - {address.country}
              </Address>
            </Temp>
            <Table>
              <Line>
                <LineTitle>Visibilidade:</LineTitle>
                <LineDecription>{info.visibility}km</LineDecription>
              </Line>
              <Line>
                <LineTitle>Umidade</LineTitle>
                <LineDecription>{info.humidity}%</LineDecription>
              </Line>
              <Line>
                <LineTitle>Temperatura min.</LineTitle>
                <LineDecription>
                  {info.temp_min && editTemp(info.temp_min)}°C
                </LineDecription>
              </Line>
              <Line>
                <LineTitle>Temperatura max.</LineTitle>
                <LineDecription>
                  {info.temp_max && editTemp(info.temp_max)}°C
                </LineDecription>
              </Line>
            </Table>
          </Details>
          <Update
            onPress={updateScreen}
            accessibilityLabel="Atualizar status"
            testID="pressButton">
            <UpdateText>Atualizar</UpdateText>
          </Update>
        </Content>
      </Background>
    </>
  );
};

export default Home;
