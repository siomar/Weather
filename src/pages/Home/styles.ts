import styled from 'styled-components/native';
import {
  SafeAreaView,
  ImageBackground,
  TouchableOpacity,
  ScrollView,
} from 'react-native';

export const Background = styled(ScrollView)`
  flex: 1;
  padding: 0px;
  position: relative;
`;

export const SafeArea = styled(SafeAreaView)`
  flex: 1;
  justify-content: center;
  align-items: center;
  padding: 0px;
  width: 100%;
  background-color: rgba(255, 255, 255, 0.8);
`;

export const Icon = styled.Image`
  border-radius: 70px;
  width: 70px;
  height: 70px;
`;

export const Title = styled.Text`
  font-size: 35px;
  line-height: 50px;
`;

export const Description = styled.Text`
  font-size: 24px;
  line-height: 26px;
  opacity: 1;
`;

export const Box = styled(ImageBackground)`
  width: 100%;
  height: 300px;
  background-color: #fff;
  margin-top: 0px;
  align-items: center;
  justify-content: center;
`;

export const Content = styled.View`
  flex: 1;
  width: 100%;
  padding: 20px 18px;
  box-sizing: border-box;
`;

export const Details = styled.View`
  margin: 14px 0px;
`;

export const Address = styled.Text`
  font-size: 18px;
  line-height: 26px;
  opacity: 1;
  align-items: center;
`;

export const Update = styled(TouchableOpacity)`
  width: 100%;
  height: 60px;
  border-radius: 10px;
  background-color: #ffcc01;
  justify-content: center;
  align-items: center;
`;

export const UpdateText = styled.Text`
  font-size: 24px;
  line-height: 26px;
  color: #fff;
`;

export const Table = styled.View`
  padding: 20px 0px;
`;

export const Line = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 15px 0px;
  border-bottom-width: 1px;
  border-bottom-color: #ddd;
`;

export const LineTitle = styled.Text`
  font-size: 20px;
  line-height: 22px;
  color: #000;
`;

export const LineDecription = styled.Text`
  font-size: 20px;
  line-height: 22px;
  color: #777;
`;

export const Loading = styled.View`
  position: absolute;
  background-color: #ffcc01;
  flex: 1;
  width: 100%;
  height: 100%;
  z-index: 200;
  justify-content: center;
  align-items: center;
`;

export const Temp = styled.View`
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
`;

export const TempTitle = styled.Text`
  font-size: 26px;
  line-height: 28px;
  color: #000;
`;
