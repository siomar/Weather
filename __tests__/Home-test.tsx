/**
 * @format
 */

import 'react-native';
import React from 'react';
import Home from '../src/pages/Home';
import {render, cleanup} from '@testing-library/react-native';

import renderer from 'react-test-renderer';
jest.mock('react-native-geolocation-service', () => jest.fn());

const tree = renderer.create(<Home />);

afterEach(cleanup);

describe('Home tests', () => {
  test('create snapshot', () => {
    expect(tree).toMatchSnapshot();
  });
  it('should find the button via testId', () => {
    const id = 'pressButton';

    const {getByTestId} = render(<Home />);

    const foundButton = getByTestId(id);

    expect(foundButton).toBeTruthy();
  });
});
