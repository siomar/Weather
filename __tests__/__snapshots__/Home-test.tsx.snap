// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home tests snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffcc01",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": "100%",
          "justifyContent": "center",
          "position": "absolute",
          "width": "100%",
          "zIndex": 200,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 24,
            "lineHeight": 26,
            "opacity": 1,
          },
        ]
      }
    >
      Carregando...
    </Text>
  </View>,
  <RCTScrollView
    nestedScrollEnabled={true}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "position": "relative",
        },
      ]
    }
  >
    <View>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "height": 300,
              "justifyContent": "center",
              "marginTop": 0,
              "width": "100%",
            },
          ]
        }
      >
        <Image
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../src/assets/bg.jpg",
            }
          }
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": 300,
                "width": "100%",
              },
              undefined,
            ]
          }
        />
        <RCTSafeAreaView
          emulateUnlessSupported={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.8)",
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "center",
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
                "width": "100%",
              },
            ]
          }
        >
          <Image
            source={
              Object {
                "uri": "https://openweathermap.org/img/wn/undefined@2x.png",
              }
            }
            style={
              Array [
                Object {
                  "borderRadius": 70,
                  "height": 70,
                  "width": 70,
                },
              ]
            }
          />
          <Text
            style={
              Array [
                Object {
                  "fontSize": 35,
                  "lineHeight": 50,
                },
              ]
            }
          >
            °C
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 24,
                  "lineHeight": 26,
                  "opacity": 1,
                },
              ]
            }
          />
        </RCTSafeAreaView>
      </View>
      <View
        style={
          Array [
            Object {
              "boxSizing": "border-box",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 20,
              "paddingLeft": 18,
              "paddingRight": 18,
              "paddingTop": 20,
              "width": "100%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginBottom": 14,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 14,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 26,
                    "lineHeight": 28,
                  },
                ]
              }
            />
            <Text
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "fontSize": 18,
                    "lineHeight": 26,
                    "opacity": 1,
                  },
                ]
              }
            >
               - 
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "paddingBottom": 20,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 20,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#ddd",
                    "borderBottomWidth": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingBottom": 15,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 15,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                Visibilidade:
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#777",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                km
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#ddd",
                    "borderBottomWidth": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingBottom": 15,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 15,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                Umidade
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#777",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                %
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#ddd",
                    "borderBottomWidth": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingBottom": 15,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 15,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                Temperatura min.
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#777",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                °C
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#ddd",
                    "borderBottomWidth": 1,
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingBottom": 15,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 15,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                Temperatura max.
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#777",
                      "fontSize": 20,
                      "lineHeight": 22,
                    },
                  ]
                }
              >
                °C
              </Text>
            </View>
          </View>
        </View>
        <View
          accessibilityLabel="Atualizar status"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffcc01",
              "borderRadius": 10,
              "height": 60,
              "justifyContent": "center",
              "opacity": 1,
              "width": "100%",
            }
          }
          testID="pressButton"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 24,
                  "lineHeight": 26,
                },
              ]
            }
          >
            Atualizar
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
